improver/metadata/check_datatypes.py

```diff
@@ -30,201 +30,180 @@
-    object_list = _construct_object_list(cube, coords)
+    # if coord, acts on coord.points
+    req_dtype = get_required_dtype(obj)
+    dtype_ok = obj.dtype == req_dtype
 
-    msg = ('{} datatype {} does not conform to expected standard ({})\n')
-    error_string = ''
-    for item in object_list:
-        # allow string-type objects
-        if item.dtype.type == np.unicode_:
-            continue
+    if isinstance(obj, iris.coords.Coord) and obj.has_bounds():
+        # check bounds - want the same dtype as the points
+        bounds_dtype_ok = obj.bounds.dtype == req_dtype
+        dtype_ok = dtype_ok and bounds_dtype_ok
+    return dtype_ok
 
-        # check numerical datatypes
-        required_dtype = _get_required_datatype(item)
-        if item.dtype.type != required_dtype:
-            error_string += msg.format(item.name(), item.dtype, required_dtype)
 
-        if (hasattr(item, "bounds") and item.bounds is not None
-                and item.bounds.dtype.type != required_dtype):
-            error_string += msg.format(
-                item.name()+' bounds', item.bounds.dtype, required_dtype)
+def get_required_units(obj):
+    """
+    Returns the mandatory units for the supplied obj. Only time coords have
+    these.
 
-    # if any data was non-compliant, raise details here
-    if error_string:
-        raise ValueError(error_string)
+    Args:
+        obj (iris.cube.Cube or iris.coords.Coord):
+            Object to be tested
 
+    Returns:
+        str or None:
+            The mandatory units corresponding to the object supplied or None
+            if there are no specific requirements for the object.
+    """
+    if _is_time_coord(obj):
+        return TIME_COORDS[obj.name()].units
+    return None
 
-def _check_units_and_dtype(obj, units, dtype):
+
+def check_units(obj):
     """
-    Check whether the units and datatype of the input object conform
-    to the standard given.
+    Checks if the supplied object complies with the relevant mandatory units.
 
     Args:
         obj (iris.cube.Cube or iris.coords.Coord):
-            Cube or coordinate to be checked
-        units (str):
-            Required units
-        dtype (type):
-            Required datatype
+            Object to be tested
 
     Returns:
         bool:
-            True if object conforms; False if not
-    """
-    if Unit(obj.units) != Unit(units):
-        return False
+            True if obj meets the mandatory units requirements or has no
+            mandatory units requirement.
 
-    if obj.dtype.type != dtype:
+    """
+    req_units = get_required_units(obj)
+    if req_units is None:
+        return True
+    if not hasattr(obj, 'units'):
         return False
-
-    return True
+    # check object and string representation to get consistent output
+    # (e.g Unit('second') == Unit('seconds'))
+    return (Unit(obj.units) == Unit(req_units) and
+            str(obj.units) == str(req_units))
```

Comment:
Apparently, we want to enforce that the unit is "seconds" and not "s". There were one or two places where this was not true.

Reply:
> I'm not sure what the string representation check is doing that's useful? As we discussed in a previous case, this will complain if my forecast period has units of 's' as they don't match 'seconds', which is an unnecessary exception as far as I'm aware.

iirc if we remove the str comparison then one of the acceptance tests will fail (which may be fine) because the incoming data has 'second' and the KGO has 'seconds'. I thought we probably wanted to make sure this was consistent, hence the check.